File: src/php/IFSC.php
<?php

namespace Razorpay\IFSC;

class IFSC
{
    protected static $data = null;
    protected static $bankNames = null;
    protected static $sublet = null;
    protected static $customSublets = null;
    protected static $customSubletPrefixes = [];

    public static function init()
    {
        if (!self::$data) {
            $contents = file_get_contents(__DIR__ . '/../IFSC.json');
            self::$data = json_decode($contents, true);
        }

        if (!self::$bankNames) {
            self::$bankNames = json_decode(file_get_contents(__DIR__ . '/../banknames.json'), true);
        }

        if (!self::$sublet) {
            self::$sublet = json_decode(file_get_contents(__DIR__ . '/../sublet.json'), true);
        }
        if (!self::$customSublets) {
            self::$customSublets = json_decode(file_get_contents(__DIR__ . '/../custom-sublets.json'), true);
            self::$customSubletPrefixes = array_keys(self::$customSublets);
        }
    }

    public static function validate(string $code)
    {
        self::init();

        if (strlen($code) !== 11) {
            return false;
        }

        if ($code[4] !== '0') {
            return false;
        }

        $bankCode   = strtoupper(substr($code, 0, 4));
        $branchCode = strtoupper(substr($code, 5));

        if (! array_key_exists($bankCode, self::$data)) {
            return false;
        }

        $list = self::$data[$bankCode];

        if (ctype_digit($branchCode)) {
            return static::lookupNumeric($list, $branchCode);
        } else {
            return static::lookupString($list, $branchCode);
        }
    }

    /**
     * Validates a given bank code
     * @param  string $bankCode 4 character bank code
     * @return boolean
     */
    public static function validateBankCode($bankCode)
    {
        return defined("Razorpay\IFSC\Bank::$bankCode");
    }

    /**
     * Returns a valid display-friendly bank name
     * @param  string $code
     * @return string or null
     */
    public static function getBankName(string $code)
    {
        self::init();

        if (self::validateBankCode($code)) {
            return self::$bankNames[$code];
        }
        else if (self::validate($code)) {
            if (isset(self::$sublet[$code])) {
                $bankCode = self::$sublet[$code];
                return self::$bankNames[$bankCode];
            }
            else {
                return self::getCustomSubletName($code) ?? self::$bankNames[substr($code, 0, 4)];
            }
        }
    }

    private static function getCustomSubletName(string $code)
    {
        foreach (self::$customSubletPrefixes as $prefix) {
            $prefixLength = strlen($prefix);
            // If the prefix matches
            if (substr($code, 0, $prefixLength) == $prefix) {
                // And the value in custom-sublets.json is a bank code
                if (strlen(self::$customSublets[$prefix]) == 4) {
                    return self::getBankName(self::$customSublets[$prefix]);
                }
                else {
                    // the value is a string that needs to be returned as-is
                    return self::$customSublets[$prefix];
                }
            }
        }
    }

    protected static function lookupNumeric(array $list, $branchCode)
    {
        $branchCode = intval($branchCode);

        if (in_array($branchCode, $list)) {
            return true;
        }

        return false;
    }

    protected static function lookupString(array $list, $branchCode)
    {
        return in_array($branchCode, $list);
    }
}

File: src/python/ifsc.py
import json
import requests
from importlib.resources import files

pkg_path = files("ifsc.data")

with open(pkg_path.joinpath("IFSC.json")) as f:
    ifscs = json.load(f)

with open(pkg_path.joinpath("banknames.json")) as f:
    bank_names = json.load(f)

with open(pkg_path.joinpath("sublet.json")) as f:
    sublets = json.load(f)

with open(pkg_path.joinpath("custom-sublets.json")) as f:
    custom_sublets = json.load(f)


class IFSC:
    def __init__(self):
        self.base_url = "https://ifsc.razorpay.com/"
        self.ifscs = ifscs
        self.bank_names = bank_names
        self.sublets = sublets
        self.custom_sublets = custom_sublets

    def validate(self, code):
        if len(code) != 11:
            return False

        if code[4] != "0":
            return False

        _bank_code = code[:4].upper()
        _branch_code = code[5:].upper()

        if not _bank_code in self.ifscs:
            return False

        if _branch_code.isnumeric():
            _branch_code = int(_branch_code)
            return _branch_code in self.ifscs[_bank_code]

        return _branch_code in self.ifscs[_bank_code]

    def get_details(self, code):
        url = self.base_url + code
        if not self.validate(code):
            raise ValueError("Invalid IFSC code")

        data = {}
        res = requests.get(url)
        if res.status_code == 200:
            data = res.json()
        elif res.status_code == 404:
            raise ValueError("Invalid IFSC code")
        else:
            raise ValueError("IFSC API returned an invalid response")

        return data

    def get_bank_name(self, code):
        if code in self.bank_names:
            return self.bank_names[code]

        if self.validate(code):
            if code in self.sublets:
                bank_code = self.sublets[code]
                return self.bank_names[bank_code]
            else:
                return self.get_custom_sublet(code)
        else:
            raise ValueError("Invalid IFSC code")

    def get_custom_sublet(self, code):
        for key, value in self.custom_sublets.items():
            if len(code) >= len(key) and code[: len(key)] == key:
                if value in self.bank_names:
                    return self.bank_names[value]
                else:
                    return value

        raise ValueError("Invalid IFSC code")

    def validate_bank_code(self, code):
        return code in self.bank_names
